### PR TITLE
JENKINS-54473 - Add support for Windows containers NetworkMode, which must be null.

### PR DESCRIPTION
**Pull request by carpnick**
https://issues.jenkins-ci.org/browse/JENKINS-54473

### CODE REVIEW

**Comment by carpnick** [comment]
Might also fix https://issues.jenkins-ci.org/browse/JENKINS-54428

File: src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSService.java [hunk 1]
```diff
@@ -232,7 +232,15 @@ TaskDefinition registerTemplate(final ECSCloud cloud, final ECSTaskTemplate temp
             LOGGER.log(Level.INFO, "Match on execution role: {0}", new Object[] {templateMatchesExistingExecutionRole});
             LOGGER.log(Level.FINE, "Match on execution role: {0}; template={1}; last={2}", new Object[] {templateMatchesExistingExecutionRole, template.getExecutionRole(), currentTaskDefinition.getExecutionRoleArn()});
 
-            templateMatchesExistingNetworkMode = StringUtils.equals(StringUtils.defaultString(template.getNetworkMode()), StringUtils.defaultString(currentTaskDefinition.getNetworkMode()));
+            
+            if(!StringUtils.isEmpty(template.getNetworkMode())){
```
Comment (stefano-pogliani) [inline src/main/java/com/cloudbees/jenkins/plugins/amazonecs/ECSService.java]: Should this be `if(StringUtils...)` without the not?
The comment suggests the then branch is for the empty string but the condition checks for not the empty string.

**Comment by pgarbe** [comment]
Is there a difference between no network mode (empty string) and `None`? Actually, I don't understand why the change is needed. Shouldn't it be enough if you specify `None` for Windows based task definitions?

**Comment by carpnick** [comment]
@pgarbe Yes there is a difference.

From [here](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#network_mode): 
`The valid values are none, bridge, awsvpc, and host`
`If the network mode is set to none, the task's containers do not have external connectivity and port mappings can't be specified in the container definition.`
`Docker for Windows uses a different network mode (known as NAT) than Docker for Linux. When you register a task definition with Windows containers, you must not specify a network mode. If you use the AWS Management Console to register a task definition with Windows containers, you must choose the default network mode.`

So in reality I think Amazon was just lazy and did not want to change their api and include another network mode of NAT after they delivered ECS.  So their answer to that was to make Null == default, which then equals NAT for windows containers.

You cant make some of this stuff up....

**Comment by pgarbe** [comment]
I think, I understand. That's really weird. What do you think, does it make sense to provide a more meaningful name (like "empty" or so) instead just the empty string? Could be confusing for non-windows users.

**Comment by carpnick** [comment]
I agree @pgarbe - I was thinking about naming it `default`, like the amazon UI does.
image

I can make the changes to make it so `default` == null for both and test both linux and windows.  Just wanted to get agreement on design before doing the testing.

**Comment by carpnick** [comment]
This is all set now @pgarbe

**Comment by pgarbe** [comment]
Published v1.18-beta-5 for testing

**Comment by carpnick** [comment]
Confirming v1.18-beta-6 is good to go. (didnt test beta-5)

**Pull request merged**